labonneboite/scripts/prepare_mailing_data.py

```diff
@@ -0,0 +1,163 @@
+# coding: utf8
+import os
+import pandas as pd
+from slugify import slugify
+
+from labonneboite.conf import settings
+from labonneboite.common import mapping as mapping_util
+from labonneboite.common import geocoding
+from labonneboite.common.search import fetch_companies
+
+INPUT_FILENAME = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'mailing_data/users_sample.csv')
+OUTPUT_FILENAME = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'mailing_data/users_sample_prepared.csv')
+
+COMMUNE_COLUMN = 'Code INSEE commune'
+ROME_1_COLUMN = 'ROME'
+ROME_2_COLUMN = 'ROME 2'
+
+COLUMNS_FOR_EACH_OFFICE = ['name', 'commune_id', 'commune_name', 'url']
+OFFICES_PER_USER = 3
+ADDITIONAL_COLUMNS = ['search_url', 'used_distance', 'used_rome_id'] + \
+    [
+        'office%s_%s' % (1+office_index, column_name)
+            for office_index in range(OFFICES_PER_USER)
+            for column_name in COLUMNS_FOR_EACH_OFFICE
+    ]
+
+# distances designed for Corsica
+DISTANCE_ATTEMPTS = [10, 20, 30]
+
+# tracking Google Analytics
+GA_TRACKING_SEARCH = 'utm_medium=email&utm_source=mailing_corse_201710&utm_campaign=lien_recherche'
+GA_TRACKING_DETAIL = 'utm_medium=email&utm_source=mailing_corse_201710&utm_campaign=lien_page_entreprise'
+
+
+def get_results(commune_id, rome_1_id, rome_2_id):
+    """
+    Try to find enough offices matching this commune_id,
+    for either of the two provided rome_ids,
+    and within as small of a distance as possible.
+
+    The result is a dictionary:
+    - offices found
+    - distance which was eventually used for the result
+    - rome_id which was eventually used for the result
+    """
+    city = geocoding.get_city_by_commune_id(commune_id)
+
+    if city is None:
+        return {'offices': [], 'distance': settings.DISTANCE_FILTER_DEFAULT, 'rome_id': rome_1_id}
+
+    latitude = city['coords']['lat']
+    longitude = city['coords']['lon']
+
+    mapper = mapping_util.Rome2NafMapper()
+
+    offices = []
+
+    # the only case where those stay None is if there is no result at all
+    used_distance = None
+    used_rome_id = None
+
+    for distance in DISTANCE_ATTEMPTS:
+        for rome_id in [rome_1_id, rome_2_id]:
+            if mapping_util.Rome2NafMapper.rome_is_valid(rome_id):
+                naf_code_list = mapper.map([rome_id])
+                offices, _, _ = fetch_companies(
+                    naf_codes=naf_code_list,
+                    rome_code=rome_id,
+                    latitude=latitude,
+                    longitude=longitude,
+                    distance=distance,
+                    from_number=1,
+                    to_number=OFFICES_PER_USER,
+                )
+                if len(offices) >= 1:
+                    used_rome_id = rome_id
+                    used_distance = distance
+                if len(offices) >= OFFICES_PER_USER:
+                    break
+        # FIXME any idea how to make this more DNRY ?
```

Comment:
Les `break` ne sont pas adaptés en fait. Il faut sortir la double boucle `for` dans une fonction séparée et remplacer les `break` par des `return`.

Reply:
peux-tu jeter un oeil à mon commit https://github.com/StartupsPoleEmploi/labonneboite/commit/a3e92bdf6b15d9e70194095a5a8f0e4a3623aab4    -   c'est bien DNRY, et je ne trouve pas que le `break` soit un problème, c'est lisible: on sort de la boucle dès qu'on a un résultat "suffisamment" bien